// @tsparticles/angular/fesm2022/tsparticles-angular.mjs
import * as i0 from '@angular/core';
import { Injectable, EventEmitter, PLATFORM_ID, Output, Input, Inject, Component, NgModule } from '@angular/core';
import { isPlatformServer } from '@angular/common';
import { tsParticles } from '@tsparticles/engine';

let initialized = false;
let initPromise;
let initCallback;
class NgParticlesService {
    get loaded() {
        return initialized;
    }
    assertInitialized() {
        if (!initialized) {
            throw new Error("NgParticlesService.init(...) must be called once before rendering ngx-particles components.");
        }
    }
    async init(particlesInit) {
        if (initialized) {
            return;
        }
        if (initPromise) {
            if (initCallback !== particlesInit) {
                throw new Error("NgParticlesService init callback must be stable across the app lifecycle.");
            }
            await initPromise;
            return;
        }
        initCallback = particlesInit;
        initPromise = (async () => {
            if (particlesInit) {
                await particlesInit(tsParticles);
            }
            await tsParticles.init();
            initialized = true;
        })().catch((error) => {
            initPromise = undefined;
            initCallback = undefined;
            initialized = false;
            throw error;
        });
        await initPromise;
    }
    async waitForInitialization() {
        await (initPromise ?? Promise.resolve());
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "22.0.4", ngImport: i0, type: NgParticlesService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
    static ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "22.0.4", ngImport: i0, type: NgParticlesService, providedIn: "root" });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "22.0.4", ngImport: i0, type: NgParticlesService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: "root",
                }]
        }] });

class NgxParticlesComponent {
    platformId;
    options;
    url;
    id = "tsparticles";
    theme;
    particlesLoaded = new EventEmitter();
    #container;
    #loadingPromise;
    #particlesService;
    constructor(platformId, particlesService) {
        this.platformId = platformId;
        this.#particlesService = particlesService;
    }
    ngAfterViewInit() {
        if (isPlatformServer(this.platformId)) {
            return;
        }
        this.#loadingPromise = this.#loadParticles();
    }
    ngOnChanges(changes) {
        if (isPlatformServer(this.platformId)) {
            return;
        }
        if (!this.#container) {
            return;
        }
        if (changes["id"] || changes["options"] || changes["url"]) {
            void this.#loadParticles();
        }
        if (changes["theme"]) {
            this.#container.loadTheme?.(changes["theme"].currentValue);
        }
    }
    ngOnDestroy() {
        this.#container?.destroy();
        this.#loadingPromise = undefined;
    }
    async #loadParticles() {
        await this.#particlesService.waitForInitialization();
        this.#particlesService.assertInitialized();
        this.#container?.destroy();
        const container = await tsParticles.load({ id: this.id, options: this.options, url: this.url });
        this.#container = container;
        if (container && this.theme) {
            container.loadTheme?.(this.theme);
        }
        this.particlesLoaded.emit(container);
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "22.0.4", ngImport: i0, type: NgxParticlesComponent, deps: [{ token: PLATFORM_ID }, { token: NgParticlesService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "22.0.4", type: NgxParticlesComponent, isStandalone: false, selector: "ngx-particles", inputs: { options: "options", url: "url", id: "id", theme: "theme" }, outputs: { particlesLoaded: "particlesLoaded" }, usesOnChanges: true, ngImport: i0, template: '<div [id]="id"></div>', isInline: true });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "22.0.4", ngImport: i0, type: NgxParticlesComponent, decorators: [{
            type: Component,
            args: [{
                    selector: "ngx-particles",
                    standalone: false,
                    template: '<div [id]="id"></div>',
                }]
        }], ctorParameters: () => [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [PLATFORM_ID]
                }] }, { type: NgParticlesService }], propDecorators: { options: [{
                type: Input
            }], url: [{
                type: Input
            }], id: [{
                type: Input
            }], theme: [{
                type: Input
            }], particlesLoaded: [{
                type: Output
            }] } });

class NgxParticlesModule {
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "22.0.4", ngImport: i0, type: NgxParticlesModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
    static ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "22.0.4", ngImport: i0, type: NgxParticlesModule, declarations: [NgxParticlesComponent], exports: [NgxParticlesComponent] });
    static ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "22.0.4", ngImport: i0, type: NgxParticlesModule, providers: [NgParticlesService] });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "22.0.4", ngImport: i0, type: NgxParticlesModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgxParticlesComponent],
                    exports: [NgxParticlesComponent],
                    providers: [NgParticlesService],
                }]
        }] });

/*
 * Public API Surface of ng-particles
 */

/**
 * Generated bundle index. Do not edit.
 */

export { NgParticlesService, NgxParticlesComponent, NgxParticlesModule };
